### PR TITLE
APEXMALHAR-2066 JdbcPolling,idempotent,partitionable

### PR DESCRIPTION
**Pull request by devtagare**
JdbcPolling,idempotent,partitionable #282

### CODE REVIEW

**Comment by devtagare** [comment]
JdbcPolling,idempotent,partitionable

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 1]
```diff
@@ -0,0 +1,592 @@
...
+   */
+  private int pollInterval;
+
+  @Min(1)
+  private int partitionCount = 1;
+  protected transient int operatorId;
+  protected transient boolean isReplayed;
+  protected transient boolean isPollable;
+  protected int batchSize;
+  protected int fetchSize;
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java]: add javadocs for properties

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 2]
```diff
@@ -0,0 +1,592 @@
...
+  private static final String user = "user";
+  private static final String password = "password";
+  /**
+   * thread to poll database
+   */
+  private transient Thread dbPoller;
+  protected transient ArrayBlockingQueue<List<T>> emitQueue;
+  protected transient PreparedStatement ps;
+  protected WindowDataManager windowManager;
+  private String emitColumnList;
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java:131]: Same as above.. Add java docs for each of the properties

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 3]
```diff
@@ -0,0 +1,592 @@
...
+  protected transient MutablePair<String, String> currentWindowRecoveryState;
+
+  /**
+   * size of the emit queue used to hold polled records before emit
+   */
+  private int queueCapacity = 4 * 1024 * 1024;
+  private transient volatile boolean execute;
+  private transient AtomicReference<Throwable> cause;
+  protected transient int spinMillis;
+  private transient OperatorContext context = null;
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java]: no need to initialize to null

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 4]
```diff
@@ -0,0 +1,592 @@
...
+  protected String tableName;
+  protected String key;
+  protected transient long currentWindowId;
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java]: What happens when checkpoint happens between Application window? currentWindowId should not be transient in that case

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 5]
```diff
@@ -0,0 +1,592 @@
...
+    super.setup(context);
+    spinMillis = context.getValue(Context.OperatorContext.SPIN_MILLIS);
+    execute = true;
+    cause = new AtomicReference<Throwable>();
+    emitQueue = new ArrayBlockingQueue<List<T>>(queueCapacity);
+    this.context = context;
+    operatorId = context.getId();
+
+    try {
+      if (ps == null) {
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java]: When is ps not null?

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 6]
```diff
@@ -0,0 +1,592 @@
...
+              java.sql.ResultSet.TYPE_FORWARD_ONLY, java.sql.ResultSet.CONCUR_READ_ONLY);
+        } else {
+          ps = store.getConnection().prepareStatement(
+              JdbcMetaDataUtility.buildPollableQuery(getTableName(), getKey(), rangeQueryPair.getKey()),
+              java.sql.ResultSet.TYPE_FORWARD_ONLY, java.sql.ResultSet.CONCUR_READ_ONLY);
+          isPollable = true;
+        }
+      }
+    } catch (SQLException e) {
+      LOG.error("Exception in initializing the range query for a given partition", e);
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java:314]: why swallow exception? Throw exception

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 7]
```diff
@@ -0,0 +1,592 @@
...
+  {
+    currentWindowId = windowId;
+
+    isReplayed = false;
+
+    if (currentWindowId <= windowManager.getLargestRecoveryWindow()) {
+      try {
+        replay(currentWindowId);
+      } catch (SQLException e) {
+        LOG.error("Exception in replayed windows");
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java]: No need to log exception

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 8]
```diff
@@ -0,0 +1,592 @@
...
+        throw new RuntimeException("saving recovery", e);
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java]: Is it "saving recovery" or replaying old data?

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 9]
```diff
@@ -0,0 +1,592 @@
...
+    this.context = context;
+    operatorId = context.getId();
+
+    try {
+      if (ps == null) {
+        //If its a range query pass upper and lower bounds
+        //If its a polling query pass only the lower bound
+        if (getRangeQueryPair().getValue() != null) {
+          ps = store.getConnection().prepareStatement(
+              JdbcMetaDataUtility.buildRangeQuery(getTableName(), getKey(), rangeQueryPair.getKey(),
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java]: Do you need to initialize "ps" here? I see that it is being done in beginWindow() again

File: library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcMetaDataUtility.java [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+public class JdbcMetaDataUtility
+{
+  private static String DB_DRIVER = "com.mysql.jdbc.Driver";
+  private static String DB_CONNECTION = "";
+  private static String DB_USER = "";
+  private static String DB_PASSWORD = "";
+  private static String TABLE_NAME = "";
+  private static String KEY_COLUMN = "";
+  private static String WHERE_CLAUSE = null;
+  private static String COLUMN_LIST = null;
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcMetaDataUtility.java:58]: why are these fields static?

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java (hunk 9 above)
Comment (devtagare) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java]: Yes.In begin window, a current partition starts with the most recent lower and upper bounds in-case of failure.In setup, the ps is initialized to the lower and upper bounds returned by the JdbcMetaDataUtility which is invoked from definePartitions().

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 11]
```diff
@@ -0,0 +1,652 @@
...
+      } else {
+        ps = store.getConnection().prepareStatement(
+            JdbcMetaDataUtility.buildPollableQuery(getTableName(), getKey(), rangeQueryPair.getKey()),
+            java.sql.ResultSet.TYPE_FORWARD_ONLY, java.sql.ResultSet.CONCUR_READ_ONLY);
+        isPollable = true;
+      }
+
+    } catch (SQLException e) {
+      LOG.error("Exception in initializing the range query for a given partition", e);
+      throw new RuntimeException();
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java]: throw new RuntimeException(e);

No need to log error

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 12]
```diff
@@ -0,0 +1,652 @@
...
+  {
+    currentWindowId = windowId;
+
+    isReplayed = false;
+
+    if (currentWindowId <= windowManager.getLargestRecoveryWindow()) {
+      try {
+        replay(currentWindowId);
+      } catch (SQLException e) {
+        LOG.error("Exception in replayed windows", e);
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java:333]: no need to log

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 13]
```diff
@@ -0,0 +1,652 @@
...
+          }
+          ps = store.getConnection().prepareStatement(
+              JdbcMetaDataUtility.buildPollableQuery(getTableName(), getKey(), bound),
+              java.sql.ResultSet.TYPE_FORWARD_ONLY, java.sql.ResultSet.CONCUR_READ_ONLY);
+          isPollable = true;
+        }
+        isReplayed = false;
+        LOG.debug("Prepared statement after re-initialization - " + ps.toString());
+      } catch (SQLException e) {
+        // TODO Auto-generated catch block
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java:361]: remove this

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 14]
```diff
@@ -0,0 +1,652 @@
...
+      throw new RuntimeException();
+    }
+
+    windowManager.setup(context);
+    LOG.debug("super setup done...");
+  }
+
+  @Override
+  public void beginWindow(long windowId)
+  {
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java:324]: begin window will not be called when checkpoint happens within Applicationwindow so how would replay happen?

May be you want to set @OperatorAnnotation(checkpointableWithinAppWindow = false) on the operator

Comment (devtagare) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java:324]: Good point.will add.

Thanks,
Dev

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 15]
```diff
@@ -0,0 +1,652 @@
...
+ * If an emitColumnList is provided, please ensure that the keyColumn is the
+ * first column in the list<br>
+ * Range queries are formed using the {@link JdbcMetaDataUtility}} Output -
+ * comma separated list of the emit columns eg columnA,columnB,columnC
+ * 
+ * @displayName Jdbc Polling Input Operator
+ * @category Input
+ * @tags database, sql, jdbc, partitionable,exactlyOnce
+ */
+public abstract class AbstractJdbcPollInputOperator<T> extends AbstractStoreInputOperator<T, JdbcStore>
```
Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java:83]: How is AbstractJdbcPollInputOperator different from AbstractJdbcInputOperator?

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java (hunk 11 above)
Comment (tweise) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java]: Should also wrap original exception, use Throwables.propagate

File: library/src/test/resources/log4j.properties [hunk 16]
```diff
@@ -39,5 +39,5 @@ log4j.appender.SYSLOG.Facility=LOCAL1
 
 log4j.logger.org=info
 #log4j.logger.org.apache.commons.beanutils=warn
-log4j.logger.com.datatorrent=debug
-log4j.logger.org.apache.apex=debug
+log4j.logger.com.datatorrent=info
```
Comment (tweise) [inline library/src/test/resources/log4j.properties]: Why this change?

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java (hunk 15 above)
Comment (devtagare) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java:83]: This operator supports polling,partitioning & idempotency.

Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java:83]: why write another and not enhance existing? Can you please point me to the discussion that happened over @dev?

Comment (pramodin) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java:83]: @gauravgopi123 @devtagare Looks like this operator is not a drop in replacement for the existing one as it makes certain assumptions about the organization of data and expects certain extra information to be made available by the implementation. In that case it would break compatibility if we try to merge it into existing. Are there use cases where the current one works correctly considering fault tolerance etc? If not we can consider deprecating it after having a discussion on the dev list.

Comment (gauravgopi123) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java:83]: @PramodSSImmaneni : I could not find these assumptions documented in JIRA. For everyone's benefit JIRA should have detailed use case information and why existing (if any) operators can't address this usecase

Comment (devtagare) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java:83]: @gauravgopi123 agreed.Apologies for the same.I will add the assumptions and the use-case.

**Comment by sandeepdeshmukh** [comment]
What are the challenges in extending AbstractJdbcInputOperator to support new features ? As AbstractJdbcInputOperator is Evolving, there is no backward compatibility issue.

**Comment by pramodin** [comment]
@sandeepdeshmukh I will let @devtagare comment but I believe this operator makes assumptions on how data is organized in the database and expects the implementations to provide that information whereas the original operator makes no such assumptions.

**Comment by gauravgopi123** [comment]
Any updates?

**Comment by devtagare** [comment]
I have updated the JIRA with assumptions for the operator - https://issues.apache.org/jira/browse/APEXMALHAR-2066.

Waiting for comments from more folks before making final changes.

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 17]
```diff
@@ -0,0 +1,652 @@
...
+  private transient OperatorContext context;
+  protected String tableName;
+  protected String key;
+  protected long currentWindowId;
+  protected KeyValPair<String, String> rangeQueryPair;
+  protected String lower;
+  protected String upper;
+  protected boolean recovered;
+  protected boolean isPolled;
+  protected String whereCondition = null;
```
Comment (bhupeshchawda) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java:119]: Can you also clarify which subset of SQL will be supported here? How complex can the where clause be?

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 18]
```diff
@@ -0,0 +1,652 @@
...
+   * Sets batchSize for number of elements in the emitQueue
+   */
+  public void setBatchSize(int batchSize)
+  {
+    this.batchSize = batchSize;
+  }
+
+  public AbstractJdbcPollInputOperator()
+  {
+    this.pollInterval = 10 * 1000;
```
Comment (bhupeshchawda) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java]: Can these be declared as private static constants as defaults?

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 19]
```diff
@@ -0,0 +1,652 @@
...
+  protected int fetchSize;
+  /**
+   * Map of windowId to <lower bound,upper bound> of the range key
+   */
+  protected transient MutablePair<String, String> currentWindowRecoveryState;
+
+  /**
+   * size of the emit queue used to hold polled records before emit
+   */
+  private int queueCapacity = 4 * 1024 * 1024;
```
Comment (bhupeshchawda) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java]: private static final int QUEUE_CAPACITY_DEFAULT?

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 20]
```diff
@@ -0,0 +1,652 @@
...
+  private transient volatile boolean execute;
+  private transient AtomicReference<Throwable> cause;
+  protected transient int spinMillis;
+  private transient OperatorContext context;
+  protected String tableName;
+  protected String key;
+  protected long currentWindowId;
+  protected KeyValPair<String, String> rangeQueryPair;
```
Comment (bhupeshchawda) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java:114]: @NotNull?

**Comment by gauravgopi123** [comment]
@devtagare : I wen through the Jira but it still doesn't mention how is this different from existing version and why can't existing version be enhanced?

File: library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcPollInputOperator.java [hunk 21]
```diff
@@ -0,0 +1,226 @@
...
+        if (isPollable) {
+          highestPolled = key.toString();
+          isPolled = true;
+        }
+      }
+      /*Flush the remaining records once the result set is over and batch-size is not reached,
+       * Dont flush if its pollable*/
+      if (!hasNext) {
+        if ((isPollable && isPolled) || !isPollable) {
+          emitQueue.add(metaList);
```
Comment (bhupeshchawda) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcPollInputOperator.java]: This might throw an exception if the queue is full. Use offer() instead?

File: library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcMetaDataUtility.java [hunk 22]
```diff
@@ -0,0 +1,344 @@
...
+    PreparedStatement preparedStatement = null;
+
+    try {
+      dbConnection = getDBConnection();
+      preparedStatement = dbConnection.prepareStatement(query);
+
+      ResultSet rs = preparedStatement.executeQuery();
+
+      while (rs.next()) {
+        rowCount = Long.parseLong(rs.getString("RowCount"));
```
Comment (bhupeshchawda) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcMetaDataUtility.java]: Suggest to use index instead of a column alias. Also seems like this is a simple count utility. Can you rename it to something like "getTupleCount()" ?

File: library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcMetaDataUtility.java [hunk 23]
```diff
@@ -0,0 +1,344 @@
...
+    PreparedStatement psUpperBound = null;
+
+    StringBuilder lowerBound = new StringBuilder();
+    StringBuilder upperBound = new StringBuilder();
+
+    KeyValPair<String, String> boundedQUeryPair = null;
+
+    try {
+      dbConnectionLower = getDBConnection();
+      dbConnectionUpper = getDBConnection();
```
Comment (bhupeshchawda) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcMetaDataUtility.java]: Do we need two database connections?

File: library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcMetaDataUtility.java [hunk 24]
```diff
@@ -0,0 +1,344 @@
...
+
+      lowerBound.append("SELECT " + KEY_COLUMN + " FROM " + TABLE_NAME);
+      upperBound.append("SELECT " + KEY_COLUMN + " FROM " + TABLE_NAME);
+
+      if (WHERE_CLAUSE != null) {
+        lowerBound.append(" WHERE " + WHERE_CLAUSE);
+        upperBound.append(" WHERE " + WHERE_CLAUSE);
+      }
+
+      lowerBound.append(" LIMIT " + (0 + lower) + ",1");
```
Comment (bhupeshchawda) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcMetaDataUtility.java:178]: LIMIT clause might not work with all databases. Need to document this.

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 25]
```diff
@@ -0,0 +1,652 @@
...
+        if (recoveredData.right.equals(rangeQueryPair.getValue()) || recoveredData.right.equals(previousUpperBound)) {
+          LOG.info("Matched so returning");
+          return;
+        }
+
+        JdbcPollInputOperator jdbcPoller = new JdbcPollInputOperator();
+        jdbcPoller.setStore(store);
+        jdbcPoller.setKey(getKey());
+        jdbcPoller.setPartitionCount(getPartitionCount());
+        jdbcPoller.setPollInterval(50);
```
Comment (bhupeshchawda) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java]: Should this be a constant? If so, can you declare this as a _DEFAULT at the top?

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 26]
```diff
@@ -0,0 +1,652 @@
...
+        jdbcPoller.setTableName(getTableName());
+        jdbcPoller.setBatchSize(getBatchSize());
+        isPollable = false;
```
Comment (bhupeshchawda) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java]: Is pollInterval needed if this is not pollable?

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 27]
```diff
@@ -0,0 +1,652 @@
...
+
+  /**
+   * Replays the tuples in sync mode for replayed windows
+   */
+  public void emitReplayedTuples(PreparedStatement ps)
+  {
+    LOG.debug("Emitting replayed statement is -" + ps.toString());
+    ResultSet rs = null;
+    try (PreparedStatement pStat = ps;) {
+      pStat.setFetchSize(10000);
```
Comment (bhupeshchawda) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java]: Can this be picked up from some property. Else declare it as a constant at the top.

File: library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcMetaDataUtility.java [hunk 28]
```diff
@@ -0,0 +1,344 @@
...
+
+  private static Connection getDBConnection()
+  {
+
+    Connection dbConnection = null;
+
+    try {
+      Class.forName(DB_DRIVER);
+    } catch (ClassNotFoundException e) {
+      LOG.error("Driver not found", e);
```
Comment (bhupeshchawda) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcMetaDataUtility.java:87]: Rethrow the exception

File: library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcMetaDataUtility.java [hunk 29]
```diff
@@ -0,0 +1,344 @@
...
+    }
+
+    try {
+      dbConnection = DriverManager.getConnection(DB_CONNECTION, DB_USER, DB_PASSWORD);
+      return dbConnection;
+    } catch (SQLException e) {
+      LOG.error("Exception in getting connection handle", e);
```
Comment (bhupeshchawda) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcMetaDataUtility.java:95]: Re throw

File: library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcMetaDataUtility.java [hunk 30]
```diff
@@ -0,0 +1,344 @@
...
+
+      ResultSet rs = preparedStatement.executeQuery();
+
+      while (rs.next()) {
+        rowCount = Long.parseLong(rs.getString("RowCount"));
+        LOG.info("# Rows - " + rowCount);
+      }
+
+    } catch (SQLException e) {
+      LOG.error("Exception in retreiving result set", e);
```
Comment (bhupeshchawda) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcMetaDataUtility.java:134]: Re throw

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java (hunk 15 above)
Comment (bhupeshchawda) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java:83]: I suggest we mark these classes as `@Evolving` since these are new and likely to change.

File: library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcMetaDataUtility.java [hunk 31]
```diff
@@ -0,0 +1,344 @@
...
+
+    try {
+      dbConnection = getDBConnection();
+      preparedStatement = dbConnection.prepareStatement(query);
+
+      ResultSet rs = preparedStatement.executeQuery();
+
+      while (rs.next()) {
+        rowCount = Long.parseLong(rs.getString("RowCount"));
+        LOG.info("# Rows - " + rowCount);
```
Comment (bhupeshchawda) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcMetaDataUtility.java]: DOn't log this.

**Comment by bhupeshchawda** [comment]
@devtagare Apart from the existing comments I am fine with the changes. 

The only major concern I have is related to the way polling is done might be dependent on a lot of things like the target database implementation and the organization of data in the database.
Request you to please mention the design of the operator (the core part on how you are planning to achieve polling functionality) on dev@apex. I guess, this will be done sooner or later as part of documentation :-) 
Apologies for the delay.

**Comment by devtagare** [comment]
@bhupeshchawda - review comments are incorporated.Could you please check.

File: library/src/test/java/com/datatorrent/lib/db/jdbc/JdbcPollerTest.java [hunk 32]
```diff
@@ -0,0 +1,246 @@
...
+    insertEventsInTable(10, 10);
+
+    AbstractJdbcPollInputOperator<Object> pollableInstance = itr.next().getPartitionedInstance();
+
+    pollableInstance.outputPort.setSink(sink1);
+
+    pollableInstance.beginWindow(1);
+    Thread.sleep(700);
+    pollableInstance.endWindow();
+    pollableInstance.emitTuples();
```
Comment (bhupeshchawda) [inline library/src/test/java/com/datatorrent/lib/db/jdbc/JdbcPollerTest.java:215]: This is emitting outside the window. Needs to be between begin and end window.

File: library/src/test/java/com/datatorrent/lib/db/jdbc/JdbcPollerTest.java [hunk 33]
```diff
@@ -0,0 +1,246 @@
...
+    }
+
+    sink1.collectedTuples.clear();
+
+    insertEventsInTable(10, 20);
+
+    pollableInstance.beginWindow(2);
+    Thread.sleep(700);
+    pollableInstance.endWindow();
+    pollableInstance.emitTuples();
```
Comment (bhupeshchawda) [inline library/src/test/java/com/datatorrent/lib/db/jdbc/JdbcPollerTest.java:232]: Same here.

**Comment by bhupeshchawda** [comment]
@devtagare I can still see a lot of open comments. Can you please address them?
Also, as I suggested, please provide the description of the design on the dev mailing list. I think this will serve the larger purpose of letting the community know about this operator and its use, in contrast to the existing Jdbc operator in Malhar.

File: library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcMetaDataUtility.java (hunk 28 above)
Comment (devtagare) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcMetaDataUtility.java:87]: Done

File: library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcMetaDataUtility.java (hunk 29 above)
Comment (devtagare) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcMetaDataUtility.java:95]: Done

File: library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcMetaDataUtility.java (hunk 24 above)
Comment (devtagare) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/JdbcMetaDataUtility.java:178]: Done

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java (hunk 20 above)
Comment (devtagare) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java:114]: At the start of the operator the rangeQueryPair would be null so a NotNull check will not work

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java (hunk 15 above)
Comment (devtagare) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java:83]: Done

**Comment by devtagare** [comment]
@bhupeshchawda - fixed the latest comments and pushed the changes.Mail sent to dev@.

File: library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java [hunk 34]
```diff
@@ -0,0 +1,656 @@
...
+      while (execute) {
+        try {
+          long startTs = System.currentTimeMillis();
+          if ((isPollable && !isPolled) || !isPollable) {
+            pollRecords(ps);
+          }
+          long endTs = System.currentTimeMillis();
+          long ioTime = endTs - startTs;
+          long sleepTime = pollInterval - ioTime;
+          LOG.debug("pollInterval = " + pollInterval + " I/O time = " + ioTime + " sleepTime = " + sleepTime);
```
Comment (tushargosavi) [inline library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java]: use {} syntax for log message, the string gets evaluated even if log level is not enabled.

**Comment by tweise** [comment]
Why would this not be part of the javadoc?

## 

sent from mobile

**Comment by bhupeshchawda** [comment]
Yes, this should be part of javadoc as well.
